Added option for chaning halt_poll_ns and for moving interrupts

dardonkov opened the pull request:
Extended the script to allow for changing of halt_poll_ns. Also gave the option the move certain interrupts to certain cpu cores. Both of these options improved performance significantly in my Ryzen 3700x VFIO setup.

dardonkov committed:
Added option for chaning halt_poll_ns and for moving interrupts

Extended the script to allow for changing of halt_poll_ns. Also gave the option the move certain interrupts to certain cpu cores. Both of these options improved performance significantly in my Ryzen 3700x VFIO setup.

ayufan commented on root/var/lib/vz/snippets/exec-cmds:
```diff
@@ -69,6 +85,14 @@ exec_cmds() {
         if_action post-start exec_cpu_taskset "$ARG1"
         ;;
 
+      "#set_halt_poll")
```
Can you add and document these options in `README`?

dardonkov replied:
Yes, I will document them and give an example.

ayufan commented on root/var/lib/vz/snippets/exec-cmds:
```diff
@@ -27,6 +27,22 @@ exec_pci_rescan() {
+}
+
+exec_assign_interrupts() {
+  ##Wait 30s for all the interrupts to show up
```
Can we add an `echo` that `wait` is happening?

dardonkov replied:
Yes, that'd be great

ayufan commented on root/var/lib/vz/snippets/exec-cmds:
```diff
@@ -27,6 +27,22 @@ exec_pci_rescan() {
+
+exec_assign_interrupts() {
+  ##Wait 30s for all the interrupts to show up
+  sleep 30s
```
I'm also curious about this, can you show what new interrupts popup, maybe less than 30s could be used?

dardonkov replied:
I tested smaller values but results were inconsistent. I think it is required for Windows to load up its GPU driver and only then all of the vfio-gpu interrupts start happening. On hardware slower than mine it might take longer, on hardware faster than mine it might be faster. 30 seconds is I think a generous value that gives enough time for Windows to start up. In my opinion it's better to wait and be of the safe side, otherwise not all interrupts are caught and moved and the optimization becomes useless. I am using irqtop to monitor the interrupts on my host machine and all other vfio interrupts show up immediately, only the GPU-related ones seem to take their time.

ayufan reviewed (changes requested):
A few requests.

dardonkov committed:
Added echo msg for the 30sec wait in exec_assign_interrupts()

ayufan commented:
@dardonkov Let me know when it is done.

dardonkov committed:
Added sections for 'set_halt_poll' and `assign_interrupts`

Added documentation and examples for 'set_halt_poll' and `assign_interrupts`. Added a small note about newer Ryzen cpu's core topology.

dardonkov committed:
Update to #assign_interrupts section

dardonkov commented:
@ayufan New options documented in README with short examples. Let me know if something needs to be changed.

ayufan commented on root/var/lib/vz/snippets/exec-cmds:
```diff
@@ -27,6 +27,22 @@ exec_pci_rescan() {
+  grep $1 /proc/interrupts | cut -d ":" -f 1 | while read -r i; do
+        echo $i
+        MASK=$2
+        echo $MASK > /proc/irq/$i/smp_affinity
```
It would be better to assign to `smp_affinity_list` that accepts a human-readable form, like: `0-11`.

dardonkov replied:
Ok, I will test it though, I agree that it's better from a usability perspective.

dardonkov replied:
Changed the script to assign to smp_affinity_list. Works as expected.

ayufan reviewed (changes requested):
This is amazing. I only added a very small usability improvement. Thank you @dardonkov!

dardonkov committed:
Changed to mark interrupts in smp_affinity_list

dardonkov committed:
Updated readme to reflect smp_affinity_list

ayufan committed:
Update README.md

ayufan committed:
Update README.md

ayufan commented:
@dardonkov Thank you for your persistence for contributing. Looks great!
